File: 1706-min-cost-to-connect-all-points/1706-min-cost-to-connect-all-points.py
import heapq

class Solution:
    def minCostConnectPoints(self, points: List[List[int]]) -> int:
        N = len(points)
        visited = set()
        min_heap = [(0, 0)]  # (cost, point_index)
        total_cost = 0

        while len(visited) < N:
            cost, i = heapq.heappop(min_heap)
            if i in visited:
                continue
            visited.add(i)
            total_cost += cost

            for j in range(N):
                if j not in visited:
                    x1, y1 = points[i]
                    x2, y2 = points[j]
                    dist = abs(x1 - x2) + abs(y1 - y2)
                    heapq.heappush(min_heap, (dist, j))

        return total_cost

File: 1706-min-cost-to-connect-all-points/1706-min-cost-to-connect-all-points.java
class Solution {
    public int minCostConnectPoints(int[][] points) {
        int n = points.length;
        Set<Integer> visited = new HashSet();
        Queue<int[]> queue = new PriorityQueue<>((a,b)->a[0]-b[0]);
        queue.offer(new int[]{0,0});  // cost, edge poin
        int minCost = 0;
        while(visited.size()<n){
            int[] arr = queue.poll();
            int i = arr[1];
            if(visited.contains(i)) continue;
            visited.add(i);
            minCost += arr[0];

            for(int j=0;j<n;j++){
                if(visited.contains(j))continue;
                int cost = Math.abs(points[i][0]-points[j][0])+Math.abs(points[i][1]-points[j][1]);
                queue.offer(new int[]{cost,j});
                queue.offer(new int[]{cost,i});
            }
        }
        return minCost;
    }
}